numpy/core/src/common/npy_cpu.c

```diff
@@ -0,0 +1,370 @@
+#include "numpy/npy_cpu.h"
+#include "numpy/npy_common.h"
+
+/******************** Private Definitions *********************/
+
+// Hold all CPU features boolen values
+static unsigned char npy__cpu_have[NPY_CPU_FEATURE_MAX];
+
+/******************** Private Declarations *********************/
+
+// Almost detect all CPU features in runtime
+static void npy__cpu_init_features(void);
+
+/******************** Public Definitions *********************/
+
+int npy_cpu_have(int feature_id)
+{
+    if (feature_id <= NPY_CPU_FEATURE_NONE || feature_id >= NPY_CPU_FEATURE_MAX)
+        return 0;
+    return npy__cpu_have[feature_id];
+}
```

Comment:
This function can be tested in the PR:
- that the set of features only appear on this CPU (easy)
- that `/proc/cpu_info` matches the set of valid features (harder)

You can write C level functions in `numpy/core/src/umath/_umath.tests.c.src and then test them from python.

Reply:
> that the set of features only appear on this CPU (easy)

Testing against `GCC` built-in functions will limit the support for X86 also `__builtin_cpu_supports` don't check OS support.

> that /proc/cpu_info matches the set of valid features (harder)

The only way I guess but it may fail on ARM if the Linux kernel is badly patched, in ` PPC64` we rely on auxiliary vector via the `LD_SHOW_AUXV` environment variable since `/proc/cpuinfo` doesn't provide information about `VSX`.